Add grant column into the users accounting calculation table

We could reuse function getGrantIfo function exported from accountsUtils.js to find grant code for each user in the table. 
Grants assigned to the user in the time of calculation would get listed. It should be mentioned in a tooltip!

@tomlebl is everything working good on your end ? the frontend container kept crashing, checked the logs and found this:  
`✘ [ERROR] Failed to resolve entry for package "ml-gsd". The package may have incorrect main/module/exports specified in its package.json. [plugin vite:dep-pre-bundle]

    node_modules/esbuild/lib/main.js:1225:21:
      1225 │         let result = await callback({
           ╵                      ^

    at packageEntryFailure (file:///app/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:46575:15)
    at resolvePackageEntry (file:///app/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:46572:3)
    at tryNodeResolve (file:///app/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:46388:16)
    at ResolveIdContext.resolveId (file:///app/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:46138:19)
    at PluginContainer.resolveId (file:///app/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:48953:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async file:///app/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:66209:15
    at async file:///app/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:46889:28
    at async requestCallbacks.on-resolve (/app/node_modules/esbuild/lib/main.js:1225:22)
    at async handleRequest (/app/node_modules/esbuild/lib/main.js:647:11)

  This error came from the "onResolve" callback registered here:

    node_modules/esbuild/lib/main.js:1150:20:
      1150 │       let promise = setup({
           ╵                     ^

    at setup (file:///app/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:46872:13)
    at handlePlugins (/app/node_modules/esbuild/lib/main.js:1150:21)
    at buildOrContextImpl (/app/node_modules/esbuild/lib/main.js:873:5)
    at Object.buildOrContext (/app/node_modules/esbuild/lib/main.js:699:5)
    at /app/node_modules/esbuild/lib/main.js:2032:68
    at new Promise (<anonymous>)
    at Object.context (/app/node_modules/esbuild/lib/main.js:2032:27)
    at Object.context (/app/node_modules/esbuild/lib/main.js:1874:58)
    at prepareEsbuildOptimizerRun (file:///app/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:50862:33)

  The plugin "vite:dep-pre-bundle" was triggered by this import

    node_modules/nmrium/lib/data/data1d/Spectrum1D/getReferenceShift.js:2:20:
      2 │ import { gsd } from 'ml-gsd';
        ╵                     ~~~~~~~~`

have you changed this after my last commit ?

--- a/nomad-front-end/src/components/AccountsComponents/AccountsTable.jsx
+++ b/nomad-front-end/src/components/AccountsComponents/AccountsTable.jsx
@@ -1,9 +1,10 @@
 import React from 'react'
-import { Table } from 'antd'
+import { Table, Tooltip } from 'antd'
 
 import classes from './AccountsTable.module.css'
 
 const AccountsTable = props => {
+  console.log(props)
   const columns = [
     {
       title: props.header,
@@ -16,6 +17,23 @@ const AccountsTable = props => {
     }
   ]
 
+  //add the grantcode if it exists
+  if (props.data[0].grantCode) {
+    columns.push({
+      title: 'Grant Code',
+      width: 35,
+      dataIndex: 'grantCode',
+      key: 'grantCode',
+      fixed: 'left',
+      align: 'center',
+      render: (text, record) => {
+        return (record.grantCode && <Tooltip title={`ID: ${record.grantCode?.grantId} \n mutiplier: ${record.grantCode?.multiplier}`}>
+          <p style={{ color: 'blue' }}>See Grants Info</p>
+        </Tooltip>)
+      }
+
+    })
+  }
   //Getting dynamic table headers from the first data object
   props.data[0].costsPerInstrument.forEach((cost, index) => {
     columns.push({
@@ -61,6 +79,7 @@ const AccountsTable = props => {
   const data = props.data.map((entry, key) => {
     const newEntry = {
       name: entry.name,
+      grantCode: entry.grantCode || undefined,
       totalCost: entry.totalCost,
       key
     }

--- a/nomad-front-end/src/components/NavBar/PageHeader/Controls/AccountingControls.jsx
+++ b/nomad-front-end/src/components/NavBar/PageHeader/Controls/AccountingControls.jsx
@@ -8,7 +8,8 @@ import { CloudDownloadOutlined } from '@ant-design/icons'
 const AccountingControls = props => {
   const { setGrantsVisible, tableData, tableHeader, accType } = props
   const standardColumns = {
-    grants: ['Grant Code', 'Description', 'Users', 'Manual Cost', 'Auto Cost', 'Total Cost [£]']
+    grants: ['Grant Code', 'Description', 'Users', 'Manual Cost', 'Auto Cost', 'Total Cost [£]'],
+    users: ['Grant Code ID', 'Grant Code Multiplier']
   }
   const columnsParser = (head, data, type) => {
     let columns = []
@@ -17,6 +18,9 @@ const AccountingControls = props => {
       columns = standardColumns.grants
     } else {
       columns = [head]
+      if (type === 'Users') {
+        columns = [...columns, ...standardColumns.users]
+      }
       let presentColumns = data[0].costsPerInstrument
       presentColumns.forEach(({ instrument }) => {
         const newColumnsToAdd = [
@@ -54,6 +58,18 @@ const AccountingControls = props => {
     } else {
       data.forEach(row => {
         let FlatRow = [row.name]
+        
+        //add the grantcodes
+        let grantCodeInfo = ['--', '--']
+        if (row.grantCode) {
+          grantCodeInfo = [row.grantCode.grantId, row.grantCode.multiplier]
+        }
+        if (type === 'Users') {
+          //only add if its users
+          FlatRow = [...FlatRow, ...grantCodeInfo]
+        }
+
+
         row.costsPerInstrument.forEach(instrumentData => {
           const { cost, expTimeAuto, expTimeClaims } = instrumentData
           FlatRow = [...FlatRow, expTimeClaims, expTimeAuto, cost]

--- a/nomad-rest-api/controllers/admin/accounts.js
+++ b/nomad-rest-api/controllers/admin/accounts.js
@@ -9,6 +9,7 @@ import Instrument from '../../models/instrument.js'
 import Grant from '../../models/grant.js'
 import {
   checkDuplicate,
+  getGrantInfo,
   getSearchParams,
   getSearchParamsClaims
 } from '../../utils/accountsUtils.js'
@@ -82,10 +83,11 @@ export async function getCosts(req, res) {
 
       await Promise.all(
         usrArray.map(async usrId => {
-          const user = await User.findById(usrId)
+          const [user, grantCode] = await Promise.all([User.findById(usrId), getGrantInfo(usrId)])
           const usrInactive = !user.isActive || user.group.toString() !== groupId
           const newEntry = {
             name: `${user.username} - ${user.fullName} ${usrInactive ? '(Inactive)' : ''}`,
+            grantCode,
             costsPerInstrument: [],
             totalCost: 0
           }

